Refactored the FitbenchmarkingParser class and added more unit tests

RabiyaF opened the pull request:
<!---

Provide a short summary of this PR in the title above.
This will be used to generate release note, so please write this in the
past tense and use language that should be understandable to a potential user.

-->

## Description of work

Fixes #1423 

Refactored the `FitbenchmarkParser` class and added more tests. This is done to avoid regression when adding the new feature.

   1) `parse`
       - [x] Refactored
       - [x] Added tests
   2) `_get_equation`
       - [x] Refactored
       - [x] Added tests
   3) `_get_starting_values`
       - [x] Refactored
       - [x] Added tests
   4) `_set_data_points`
       - [x] Refactored
       - [x] Added tests
   5) `_get_data_file`
       - [x] Refactored
       - [x] Added tests
   6) `_get_data_problem_entries`
       - [x] Refactored
       - [x] Added tests
   7) `_parse_function`
       - [x] Refactored
       - [x] Added tests
   8) `_parse_jac_function`
       - [x] Refactored
       - [x] Added tests
   9) `_dense_jacobian`
       - [x] Refactored
       - [x] Added tests
   10) `_sparse_jacobian`
       - [x] Refactored
       - [x] Added tests
   11) `_get_jacobian`
       - [x] Refactored
       - [x] Added tests
   12) `_parse_single_function`
       - [x] Refactored
       - [x] Added tests
   13) `_parse_parens`
       - [x] Refactored
       - [x] Added tests
   14) `_parse_function_value`
       - [x] Refactored
       - [x] Added tests
   15) `_get_data_points`
       - [x] Refactored
       - [x] Added tests
   16) `_parse_range`
       - [x] Refactored
       - [x] Added tests
   17) `_find_first_line`
       - [x] Refactored
       - [x] Added tests
   18) `_get_column_data`
       - [x] Refactored
       - [x] Added tests
   19) `_is_multifit`
       - [x] Refactored
       - [x] Added tests
   20) `_get_plot_scale`
       - [x] Refactored
       - [x] Added tests

<!---

Describe your changes and why you're making them.

-->


## Testing Instructions

<!---

Please give any specific testing instructions to the reviewer here.

-->

1. Verify that the test coverage of `fitbenchmark_parser.py` is **100%**.
<img width="476" alt="image" src="https://github.com/user-attachments/assets/17f4d3d7-7fba-4e59-aa1d-d6e82d102e62" />

2. Verify that coveralls reports a coverage increase.
<img width="220" alt="image" src="https://github.com/user-attachments/assets/b2477f69-78df-4b24-bc05-78b0099ab67a" />

3. Run the tests using `pytest fitbenchmarking/parsing/tests/test_fitbenchmark_parser.py`
4. Run different cli commands to verify that there is no regression in performance.


## Checklist

<!---

This checklist is mostly useful as a reminder of small things that can easily be

forgotten – it is meant as a helpful tool rather than hoops to jump through.

Put an `x` in all the items that apply, make notes next to any that haven't been

addressed, and remove any items that are not relevant to this PR.

-->

- [x] The title is of a format appropriate for a line in future release notes.
- [x] I have added the appropriate tags to the PR.
- [x] My PR represents one logical piece of work.
- [x] My PR fully fixes the issue linked. If new issues have been created, what are they?
- [x] I have added the appropriate tests to cover code that has been added.
- [x] I have updated the documentation in the relevant places to cover the changes.

jess-farmer reviewed (approved):
Many thanks for this refactor @RabiyaF, all looks good to me! 💯

jess-farmer commented:
@RabiyaF The windows integration tests are failing due to `PosixPath` giving the following error: `NotImplementedError: cannot instantiate 'PosixPath' on your system`. For testing on Windows, I think we need to use [`WindowsPath`](https://docs.python.org/3/library/pathlib.html#pathlib.WindowsPath) instead?

RabiyaF commented:
I will rebase and have a look at this.

RabiyaF commented:
I removed more dead code from the `parse` method. I am pretty confident that this PR should not break anything. The only thing I changed logically was the following condition because I think it should be `> 1`.

<img width="909" alt="image" src="https://github.com/user-attachments/assets/965db9c2-e160-48ce-93f1-0bcc1792fc85" />
